core/src/main/scala/org/apache/spark/scheduler/BlacklistStrategy.scala

```diff
@@ -0,0 +1,145 @@
+      true
+    }
+  }
+}
+
+/**
+ * This strategy is simply based on given threshold and is taskId unrelated. An executor will be
+ * in blacklist, if it failed more than "maxFailureTaskNumber" times. A node will be in blacklist,
+ * if there are more than "maxBlackExecutorNumber" executors on it in executor blacklist.
+ *
+ * In this case, provided taskId will be ignored. The benefit for taskId unrelated strategy is that
+ * different taskSets can learn experience from other taskSet to avoid allocating tasks on
+ * problematic executors.
+ */
+private[scheduler] class SimpleStrategy(
+    maxFailureTaskNumber: Int,
+    maxBlackExecutorNumber: Int,
+    val expireTimeInMilliseconds: Long
+  )extends BlacklistStrategy {
+
+  private def getSelectedExecutorMap(
+      executorIdToFailureStatus: mutable.HashMap[String, FailureStatus]) = {
+    executorIdToFailureStatus.filter{
+      case (id, failureStatus) => failureStatus.totalNumFailures > maxFailureTaskNumber
+    }
+  }
+
+  // As this is a taskId unrelated strategy, the input taskId will be ignored
+  def getExecutorBlacklist(
+      executorIdToFailureStatus: mutable.HashMap[String, FailureStatus],
+      taskIndex: Int, clock: Clock): Set[String] = {
+    getSelectedExecutorMap(executorIdToFailureStatus).keys.toSet
+  }
+
+  def getNodeBlacklist(
+      executorIdToFailureStatus: mutable.HashMap[String, FailureStatus]): Set[String] = {
+    getSelectedExecutorMap(executorIdToFailureStatus)
+      .groupBy{case (id, failureStatus) => failureStatus.host}
+      .filter {case (host, executorIdToFailureStatus) =>
+        executorIdToFailureStatus.size > maxBlackExecutorNumber}
+      .keys.toSet
+  }
+}
+
+/**
+ * This strategy is applied as default to keep the same semantics as original. It's an taskId
+ * related strategy. If an executor failed running "task A", then we think this executor is
+ * blacked for "task A". And we think the executor is still healthy for other task. node blacklist
+ * is always empty.
+ *
+ * It was the standard behavior before spark 1.6
+ */
+private[scheduler] class DefaultStrategy(
+    val expireTimeInMilliseconds: Long) extends BlacklistStrategy {
+  def getExecutorBlacklist(
+      executorIdToFailureStatus: mutable.HashMap[String, FailureStatus],
+      taskIndex: Int, clock: Clock): Set[String] = {
+    executorIdToFailureStatus.filter{
+      case (_, failureStatus) => failureStatus.numFailuresPerTask.keySet.contains(taskIndex) &&
+        clock.getTimeMillis() - failureStatus.updatedTime < expireTimeInMilliseconds
+    }.keys.toSet
+  }
+
+  def getNodeBlacklist(
+      executorIdToFailureStatus: mutable.HashMap[String, FailureStatus]): Set[String] =
+        Set.empty[String]
+}
+
+/**
+ * Create BlacklistStrategy instance according to SparkConf
+ */
+private[scheduler] object BlacklistStrategy {
+  def apply(sparkConf: SparkConf): BlacklistStrategy = {
+    val timeout = sparkConf.getLong("spark.scheduler.executorTaskBlacklistTime", 0L)
```

Comment:
use *getTimeAsMs* instead and I think time unit could be automatically converted "s <--> ms <--> us " by this method, so it's not necessary to divide 1000 here.